Базовый вариант сервера

kaktycc opened the pull request:
Добавлена обработка, шифрование данных. Создание токена, обработка и отправка JSON.
Добавлена основа для работы с вебсокетами. Добавлена реализация
получения списка пользователей из бд с проверкой токена

kaktycc committed:
Базовый вариант сервера. Добавлена обработка аутентификационных
данных, шифрование данных. Создание токена, обработка и отправка JSON.
Добавлена основа для работы с вебсокетами. Добавлена реализация
получения списка пользователей из бд с проверкой токена

AlwxSin commented:
@kaktycc привет.
Сразу и на будущее. Очень плохо делать все одним коммитом и в одном файле.
Про коммиты:
* В случае чего, невозможно откатить какое-то атомарное изменение.
* Сложно просматривать изменения. Если коммит атомарный, то можно быстро его обсудить и, в случае чего, что-нибудь поменять.

Про один файл:
Если код разнесен по логически выбранным файлам/модулям, то его очень легко переиспользовать.
Сейчас прикинь, что можно вынести из этого файла в вспомогательные (например работа с БД или аутентификация).

AlwxSin commented on server.py:
```diff
@@ -0,0 +1,211 @@
+    def __init__(self):
+        self.webSocketsPool = []
+        handlers = [
+            (r'/v1', MainHandler),
```
Подобные хэндлеры принято организовывать по раздельным модулям. Это к разносу кода по файлам/модулям.
А в главном файле импортировать.

AlwxSin commented on server.py:
```diff
@@ -0,0 +1,211 @@
+            (r'/v1/test/', TestHandler)
+
+        ]
+        settings = dict()
```
Если оставил пустой `settings` на будущее, стоит написать коммент, для чего он может понадобится.

AlwxSin commented on server.py:
```diff
@@ -0,0 +1,211 @@
+        tornado.web.Application.__init__(self, handlers, **settings)
+
+        # bd connection
+        self.db = create_engine('postgresql://postgres:123@localhost:5432/messengerbase')
```
Сейчас любой, кто попытается запустить код (или при деплое) вынужден будет создавать у себя в постгресе базу с определенным именем, пользователя и пароль к нему `123`.
Хорошая практика - вынести подобное в переменные окружения.

AlwxSin commented on server.py:
```diff
@@ -0,0 +1,211 @@
+                # print(self.request.body)
+                arg = self.request.body
+                arg = arg.decode()
+                arg = arg.replace("'", '"')
```
Очень странная конструкция.
Зачем понадобилось заменять одинарные кавычки на двойные? Почему с фронта может прийти такая конструкция?
Если она все-таки придет, не лучше ли сказать фронту, что у него плохие данные?

AlwxSin commented on server.py:
```diff
@@ -0,0 +1,211 @@
+        incoming_arg_1 = self.json_data['token']
+        incoming_arg_2 = self.json_data['name']
+
+        print(incoming_arg_1)
```
Оставлять принты - не очень хорошая идея.
Для дебага локально еще нормально, но в репозиторий пушить, чтобы все видели - не очень =)

AlwxSin commented on server.py:
```diff
@@ -0,0 +1,211 @@
+        passwd = self.json_data['password']
+
+        credpass = hashlib.md5(passwd.encode('utf8')).hexdigest()
+        if login == 'admin' and credpass == hashlib.md5('admin'.encode('utf8')).hexdigest():
```
Это нужно срочно править.
Чтобы в БД юзеры хранились, а пароли захэшированы и с солью.

AlwxSin commented on server.py:
```diff
@@ -0,0 +1,211 @@
+
+class UsersHandler(JsonHandler):
+    def get(self):
+        # token сверить с БД
```
Комментарий относится к тому, что надо в БУДУЩЕМ сделать?
Если да, то надо вставить префикс `# TODO do something in future`.
Такие префиксы распознаются редакторами.

AlwxSin commented on server.py:
```diff
@@ -0,0 +1,211 @@
+    def get(self):
+        # token сверить с БД
+        # если токен есть вывести список всех юзеров
+        result_set = self.db.execute("SELECT * FROM users")
```
Не видно схемы БД, но если в базе будут храниться пароли и/или любая другая чувствительная информация, то этот запрос всю ее и достанет.

AlwxSin commented on server.py:
```diff
@@ -0,0 +1,211 @@
+        # print(result_set)
+
+
+class UserAddHandler(JsonHandler):
```
Добавь, пожалуйста, докстринг с описанием того, что делает этот хэндлер.
А то непонятно, вроде должен добавлять юзера, но по коду он только достает данные из базы.

AlwxSin commented on server.py:
```diff
@@ -0,0 +1,211 @@
+            (r'/v1/auth', AuthHandler),
+            (r'/v1/users', UsersHandler),
+            (r'/v1/users/add', UserAddHandler),
+            (r'/v1/wsocket/', WebSocketHandler),
```
Не очень удачное сокращение.
Принято либо  `ws`, либо полностью.

AlwxSin reviewed (commented):
Структура более менее понятна.
Но ты закоммитил много болванок.
В таких случаях лучше либо вообще не коммитить, либо, что в данном случае нужно, оставлять подробные комменты с префиксом `# TODO`

kaktycc commented:
спасибо за комментарии, поправил недочеты

AlwxSin commented:
@kaktycc А зачем ты закрыл PR? Можно в рамках него коммитить.

kaktycc committed:
Разделил на классы в отдельных файлах.
Добавил соль и хэш в аутентификации

freedomexists committed:
added bd-tools

AlwxSin commented on handlers/authhandler.py:
```diff
@@ -0,0 +1,42 @@
+    def create_sha(self, string):
+        return hashlib.sha256(string.encode() + salt.encode()).hexdigest()
+
+    def post(self):
```
Смотри какая тут сложилась ситуация.
У тебя такой алгоритм:
1) Посмотреть есть ли юзер с подобным логином в БД, если нет - вернуть ошибку.
2) Посмотреть совпадает ли хэшированный пароль с тем, что в БД, если нет - вернуть ошибку.
3) Залогинить пользователя.

А почему бы не совместить первые два пункта. Ты можешь за один запрос (а уменьшить количество запросов к БД очень важно) проверить и пользователя и пароль.

AlwxSin commented on handlers/authhandler.py:
```diff
@@ -0,0 +1,42 @@
+        login = self.json_data['user']
+        passwd = self.json_data['password']
+        passwd = self.create_sha(passwd)
+        query = "SELECT COUNT(login) FROM users WHERE login = '" + self.json_data['user'] + "'"
```
Прочитай про SQL-инъекции.
В твоем коде сейчас ее нельзя сделать. Но недалеко от того.
Нужно обязательно санитайзить входные данные, которые ты подставляешь в запрос к БД.

kaktycc replied:
что значит санитайзить? убирать лишние символы?

AlwxSin replied:
https://blog.sqreen.io/preventing-sql-injections-in-python/

AlwxSin commented on handlers/jsonhandler.py:
```diff
@@ -0,0 +1,42 @@
+import tornado.escape
```
Так как этот файл не содержит непосредственно хэндлеров, которые обрабатывают запросы, то может возникнуть некая путаница.
Возможно стоит переименовать его в `utils.py` или что-то подобное?

AlwxSin commented on handlers/jsonhandler.py:
```diff
@@ -0,0 +1,42 @@
+        if self.request.body:
+            try:
+                # print(self.request.body)
+                # для дебага через curl
```
Я правильно понял, что ты меняешь кавычки только для дебага через curl?
Если так, то плохая практика. Твои коллеги могут вместо `curl` использовать, например Postman, и их будет очень смущать подобное поведение.

AlwxSin reviewed (commented):
Отлично, теперь код намного понятнее организован и написан.

freedomexists committed:
Merge branch 'develop' into develop

kaktycc committed:
Merge pull request #1 from freedomexists/develop

added bd-tools

AlwxSin commented on .idea/vcs.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Обязательно следи за тем, что ты добавляешь в гит. Особенно, если ты работаешь с гитом вне командной строки.
Плюс, есть такая замечательная вещь, как `.gitignore`

kaktycc committed:
Добавлен gitignore
Изменено название папки для работы с БД
Вынесено в отдельный модуль подключение к БД

kaktycc committed:
Переименован файл для работы с JSON,
Добавлена проверка токена

kaktycc committed:
Добавлено поле email

kaktycc committed:
Добавлены библиотеки для работы с БД и JSON

kaktycc committed:
Добавлена проверка токена при любом методе,
Добавлена функция вывода данных о пользователях из БД,
Добавлена функция записи пользователя в БД

kaktycc committed:
Добавлена генерация токена

kaktycc committed:
Изменена колонка таблицы с content на message,
Добавлен адрес для /chats/add

kaktycc committed:
Добавлен вывод всех чатов из БД,
Создание нового чата

kaktycc committed:
Добавлен геттер для элементов из БД

freedomexists committed:
reworked bd queries

freedomexists committed:
Merge pull request #1 from kaktycc/develop

Develop

freedomexists committed:
Merge remote-tracking branch 'origin/develop'

# Conflicts:
#	database_tools/init.py
#	handlers/usershandler.py

freedomexists committed:
reworked bd queries

kaktycc committed:
Merge pull request #2 from freedomexists/develop

Develop

kaktycc committed:
изменен userhandler

freedomexists committed:
Merge pull request #2 from kaktycc/develop

Develop

kaktycc committed:
добавил в гитигнор

freedomexists committed:
added table contacts

freedomexists committed:
Изменил _tocken_check чтобы возвращал uid, вместо тру, стобы уменьшить кол-во запросов к бд.
Добавил обработку запроса на внесение в список контактов.

freedomexists committed:
Правки к предыдущему коммиту.
Заготовка метода под обработку запроса на получения статуса пользователей из списка контактов

kaktycc committed:
Merge pull request #3 from freedomexists/develop

Develop

kaktycc committed:
Исправлен импорт модуля, post поменян на put
